File: src/main/kotlin/ca/yakov/aoc/aoc2016/day13.kt
package ca.yakov.aoc.aoc2016

import ca.yakov.aoc.Solution
import kotlin.math.abs

private val start = Coord(1, 1)
private val goal = Coord(31, 39)
private val spaces = mutableMapOf<Coord, Boolean>()

private data class Coord(val x: Int, val y: Int)
private class Node(val coord: Coord, val dist: Int)
private class Stats(coord: Coord, val gScore: Int) {
    val fScore = gScore + abs(coord.x - goal.x) + abs(coord.y - goal.y)
}

fun run13(input: String): Solution {
    val inputInt = input.trim().toInt()
    val closedSet = mutableSetOf(start)
    val openMap = mutableMapOf<Coord, Stats>()
    val queue = mutableListOf(Node(start, 0))

    while (queue.isNotEmpty()) {
        val node = queue.removeFirst()
        val coord = node.coord
        closedSet.add(coord)
        if (node.dist == 50) {
            openMap += coord to Stats(coord, 50)
            continue
        }

        listOf(
            Coord(coord.x - 1, coord.y),
            Coord(coord.x + 1, coord.y),
            Coord(coord.x, coord.y - 1),
            Coord(coord.x, coord.y + 1)
        )
            .filter { it !in closedSet && it.x >= 0 && it.y >= 0 && isSpace(it, inputInt) }
            .forEach {
                closedSet += it
                queue += Node(it, node.dist + 1)
            }
    }

    val visited = closedSet.size.toULong()
    val leastSteps = aStar(openMap, closedSet, inputInt)
    return Solution(leastSteps, visited)
}

private fun aStar(openMap: MutableMap<Coord, Stats>, closedSet: MutableSet<Coord>, input: Int): ULong {
    while (true) {
        val node = openMap.minByOrNull { (_, v) -> v.fScore }!!
        val coord = node.key
        if (coord == goal) {
            return node.value.gScore.toULong()
        }

        val tentativeGscore = node.value.gScore + 1
        openMap -= coord
        closedSet += coord

        listOf(
            Coord(coord.x - 1, coord.y),
            Coord(coord.x + 1, coord.y),
            Coord(coord.x, coord.y - 1),
            Coord(coord.x, coord.y + 1)
        )
            .filter {
                it !in closedSet && it.x >= 0 && it.y >= 0 &&
                        openMap[it]?.let { it.gScore > tentativeGscore } != false &&
                        isSpace(it, input)
            }
            .forEach {
                openMap += it to Stats(it, tentativeGscore)
            }
    }
}

private fun isSpace(coord: Coord, input: Int) =
    spaces.getOrPut(coord) {
        (coord.x * coord.x + 3 * coord.x + 2 * coord.x * coord.y + coord.y + coord.y * coord.y + input)
            .countOneBits() % 2 == 0
    }

File: src/main/nim/aoc/2016/day13.nim
import std/bitops, std/sets, std/strutils, std/tables, zero_functional, ../aoc

type
  Coord = tuple[x: int, y: int]
  Node = tuple[coord: Coord, dist: int]
  Stats = tuple[coord: Coord, gScore: int]

const
  start: Coord = (1, 1)
  goal: Coord = (31, 39)

proc fScore(stats: Stats): int = 
  stats.gScore + abs(stats.coord.x - goal.x) + abs(stats.coord.y - goal.y)

proc `<`(a, b: (Coord, Stats)): bool = a[1].fScore() < b[1].fScore()

proc isSpace(coord: Coord, input: int, spaces: var Table[Coord, bool]): bool =
  if coord notin spaces:
    spaces[coord] =
      (coord.x * coord.x + 3 * coord.x + 2 * coord.x * coord.y + coord.y + coord.y * coord.y + input)
        .countSetBits() %% 2 == 0
  spaces[coord]

proc aStar(openMap: var Table[Coord, Stats], closedSet: var HashSet[Coord], input: int, spaces: var Table[Coord, bool]): int =
  while true:
    let node = block: # hacky stuff to get around the compiler complaining
      let openMapCopy = openMap
      openMapCopy.pairs --> min()
    let coord = node[0]
    if coord == goal:
      return node[1].gScore

    let tentativeGscore = node[1].gScore + 1
    openMap.del(coord)
    closedSet.incl(coord)

    @[
      (x: coord.x - 1, y: coord.y),
      (x: coord.x + 1, y: coord.y),
      (x: coord.x, y: coord.y - 1),
      (x: coord.x, y: coord.y + 1),
    ] -->
      filter(
        it notin closedSet and it.x >= 0 and it.y >= 0 and
          (it notin openMap or openMap[it].gScore > tentativeGscore) and
          it.isSpace(input, spaces)
      ) -->
      foreach(
        block: openMap[it] = (it, tentativeGscore)
      )

proc day13*(file: string): Solution =
  let inputInt = file.parseInt()
  var closedSet = @[start].toHashSet()
  var openMap: Table[Coord, Stats]
  var queue: seq[Node] = @[(coord: start, dist: 0)]
  var spaces: Table[Coord, bool]

  while queue.len() != 0:
    let node = queue[0]
    queue.delete(0)
    let coord = node.coord
    closedSet.incl(coord)
    if node.dist == 50:
      openMap[coord] = (coord: coord, gscore: 50)
      continue

    @[
      (x: coord.x - 1, y: coord.y),
      (x: coord.x + 1, y: coord.y),
      (x: coord.x, y: coord.y - 1),
      (x: coord.x, y: coord.y + 1)
    ] -->
      filter(it notin closedSet and it.x >= 0 and it.y >= 0 and it.isSpace(inputInt, spaces)) -->
      foreach(block:
        closedSet.incl(it)
        queue &= (coord: it, dist: node.dist + 1)
      )

  let visited = closedSet.len()
  let leastSteps = aStar(openMap, closedSet, inputInt, spaces)
  Solution(s1: $leastSteps, s2: $visited)
